### PR TITLE
[EDA] Refactor: CON_PrimaryLanguage_TDTM

### PR DESCRIPTION
**Pull request by m-kinkade**
# Critical Changes

# Changes
## General Updates
To improve readability, modularity, and performance, we've refactored the CON_PrimaryLanguage_TDTM class. The functionality of this class hasn't changed.

# Issues Closed

# New Metadata

# Deleted Metadata

# Testing Notes
https://salesforce.quip.com/g3GVAOxfWsdH#WOWACA1Jzbj

### CODE REVIEW

File: src/classes/SRVC_ContactPrimaryLanguage.cls [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+    * @description Empty constructor is not public for Singleton Pattern.
+    **********************************************************************/
+    protected SRVC_ContactPrimaryLanguage() {}
+
+    /*****************************************************************************
+    * @description Static method to get the current instance for Singleton pattern.
+    * @return The instance of SRVC_ContactPrimaryLanguage.
+    ******************************************************************************/
+    public static SRVC_ContactPrimaryLanguage getInstance() {
+        if (instance == NULL) {
```
Comment (MatthewBlanski) [inline src/classes/SRVC_ContactPrimaryLanguage.cls:56]: I realize I've let this slide in the past. If we can include the class name on any references to static methods or members, that would be ideal.

File: src/classes/SRVC_ContactPrimaryLanguage.cls [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+
+        return conLangsToProcess;
+    }
+
+    /************************************************************************************************************
+    * @description Value object specifying a list of new Contact Language records to be inserted 
+    * and a list of existing Contact Language records to be updated.
+    ************************************************************************************************************/
+    public class ContactLanguagesToProcess{
+        public List<Contact_Language__c> contactLanguagesToInsertList { get; private set; }
```
Comment (MatthewBlanski) [inline src/classes/SRVC_ContactPrimaryLanguage.cls:214]: @m-kinkade This is an extremely efficient way to handle the situation with the access controls we desired.

File: src/classes/SRVC_ContactPrimaryLanguage_TEST.cls [hunk 3]
```diff
@@ -0,0 +1,637 @@
...
+*/
+@isTest
+private class SRVC_ContactPrimaryLanguage_TEST {
+
+    /**************************************************************************************************************************
+    * @description Test method to verify that the getInstance method returns the existing instance of the 
+    * SRVC_ContactPrimaryLanguage class or a new one if one doesn't already exist.
+    ***************************************************************************************************************************/
+    @isTest
+    private static void getInstancePositive() {
```
Comment (MatthewBlanski) [inline src/classes/SRVC_ContactPrimaryLanguage_TEST.cls:45]: @m-kinkade As the addendum to my comment about getting the same instance, this is the appropriate location to make certain that this is working as a singleton.

File: src/classes/SRVC_ContactPrimaryLanguage_TEST.cls [hunk 4]
```diff
@@ -0,0 +1,637 @@
...
+        System.assertEquals(true, conLangsToSetNonPrimary.isEmpty(), 'There should be no Contact Languages to be updated to non-primary status.');
+    }
+
+    /**************************************************************************************************************************
+    * @description Test method to verify that the getDefaultFluencyFromSettings correctly returns the value set as the 
+    * Default_Contact_Language_Fluency__c.
+    ***************************************************************************************************************************/
+    @isTest 
+    private static void getDefaultFluencyFromSettingsPositive(){
+        Hierarchy_Settings__c hs = UTIL_CustomSettingsFacade.getSettingsForTests(new Hierarchy_Settings__c
```
Comment (MatthewBlanski) [inline src/classes/SRVC_ContactPrimaryLanguage_TEST.cls:270]: @m-kinkade This is a good example of where using getSettingsForTests is appropriate to test the veracity of that method.

File: src/classes/SRVC_ContactPrimaryLanguage_TEST.cls [hunk 5]
```diff
@@ -0,0 +1,637 @@
...
+
+        System.assertEquals(true, String.isBlank(fluencySetting), 'Default fluency setting should be null.');
+    }
+
+    /**************************************************************************************************************************
+    * @description Test method to verify that the getDefaultFluencyFromSettings correctly returns a blank value when 
+    * the value of Default_Contact_Language_Fluency__c is blank.
+    ***************************************************************************************************************************/
+    @isTest 
+    private static void getDefaultFluencyFromSettingsBlank(){
```
Comment (MatthewBlanski) [inline src/classes/SRVC_ContactPrimaryLanguage_TEST.cls:301]: @m-kinkade What value does this test add to our suite in light of the other tests?

File: src/classes/SRVC_ContactPrimaryLanguage.cls [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+    }
+
+    /************************************************************************************************************
+    * @description Determines if changes to the Primary Language on Contact records are associated with existing 
+    * Contact Language records or if they require the creation of new Contact Language records.             
+    * @param  newContactsById A map of updated contact records by Id.
+    * @param  oldContactsById A map containing the previous state of the updated contact records by Id.
+    * @return A ContactLanguagesToProcess value object containing a list of new Contact Language records to insert and a list of existing Contact Language records to update.
+    ************************************************************************************************************/
+    @TestVisible
```
Comment (skosgi1) [inline src/classes/SRVC_ContactPrimaryLanguage.cls:94]: minor, not needed for public methods.

File: src/classes/CON_PrimaryLanguage_TDTM.cls [hunk 7]
```diff
@@ -27,139 +27,169 @@
...
 
-                    primaryLanguageByContact.put(newContact.Id, newContact.Primary_Language__c);
-                }
-            }
+    /*******************************************************************************************************
+     * @description Handles processing required for an After Insert trigger.
+     * @param newSObjectsList The list of SObjects inserted.
+     *******************************************************************************************************/ 
+    @TestVisible
+    private void handleAfterInsert(List<SObject> newSObjectsList){
```
[the remark this answers is not in the document]
Comment (MatthewBlanski) [reply src/classes/CON_PrimaryLanguage_TDTM.cls:74]: The choice is deliberate as it intentionally borrows from standard naming for event-based processing, which is closer to what we are now implementing.

**Review by MatthewBlanski — APPROVED** [review]
These changes are functional and effective.

File: src/classes/CON_PrimaryLanguage_TDTM.cls (hunk 7 above)
[the remark this answers is not in the document]
Comment (xinnanlisf) [reply src/classes/CON_PrimaryLanguage_TDTM.cls:74]: Does this also mean we need the rename those methods in CON_DoNotContact_TDTM?

[the remark this answers is not in the document]
Comment (xinnanlisf) [reply src/classes/CON_PrimaryLanguage_TDTM.cls:74]: https://github.com/SalesforceFoundation/EDA/blob/22ab77bd8f0d56e41c9c6361e01abf7203e2142d/src/classes/CON_DoNotContact_TDTM.cls#L92

**Review by lifewithryan — APPROVED** [review]
Looks like this was approved before and the new changes are for the package.xml conflicts. approving for merge freeze.

**Comment by spinkelman** [comment]
QA Passes - Merging
Will close WI

**Pull request merged**